sql/connect/server/src/main/scala/org/apache/spark/sql/connect/ml/MLUtils.scala

```diff
@@ -50,8 +51,18 @@ private[ml] object MLUtils {
   private def loadOperators(mlCls: Class[_]): Map[String, Class[_]] = {
     val loader = Utils.getContextOrSparkClassLoader
     val serviceLoader = ServiceLoader.load(mlCls, loader)
-    val providers = serviceLoader.asScala.toList
-    providers.map(est => est.getClass.getName -> est.getClass).toMap
+    // Instead of using the iterator, we use the "stream()" method that allows
+    // to iterate over a collection of providers that do not instantiate the class
+    // directly. Since there is no good way to convert a Java stream to a Scala stream,
+    // we collect the Java stream to a Java map and then convert it to a Scala map.
+    serviceLoader
+      .stream()
+      .collect(
+        Collectors.toMap(
+          (est: ServiceLoader.Provider[_]) => est.`type`().getName,
+          (est: ServiceLoader.Provider[_]) => est.`type`()))
+      .asScala
```

Comment:
Hm, sorry if I got this wrong, but we still convert the stream to the map so all the instances will be created.

Reply:
This will only be useful if we do some filtering, etc. between `stream()` and `collect()`.

Reply:
@HyukjinKwon its not about the end result but the types of the iterator. 

The general iterstor over the service loader returns an iterator over the instances. The stream method returns an iterator over a stream of providers that have not instantiated the classes. 

See https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/ServiceLoader.html#stream()

Reply:
ah okie